From bd400c38d84713bb36a6003bf4dfa3a198917d0c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 22 Feb 2022 10:43:06 +0100
Subject: [PATCH 28/35] winegstreamer: Check audio/video info from caps in
 wg_pool_create.

For: Call of Duty III, Mortal Kombat 11, Shadow Warrior 2,
Yakuza 4 Remastered, Hard Reset Redux.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=45988
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=47084
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=49715
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=52183
CW-Bug-Id: #16839
CW-Bug-Id: #18678
CW-Bug-Id: #19362
---
 dlls/winegstreamer/unix_private.h |  2 +-
 dlls/winegstreamer/wg_pool.c      | 20 +++++++++++++++++++-
 dlls/winegstreamer/wg_transform.c |  2 +-
 3 files changed, 21 insertions(+), 3 deletions(-)

diff --git a/dlls/winegstreamer/unix_private.h b/dlls/winegstreamer/unix_private.h
index 11111111111..11111111111 100644
--- a/dlls/winegstreamer/unix_private.h
+++ b/dlls/winegstreamer/unix_private.h
@@ -39,7 +39,7 @@ extern NTSTATUS wg_transform_destroy(void *args) DECLSPEC_HIDDEN;
 extern NTSTATUS wg_transform_push_data(void *args) DECLSPEC_HIDDEN;
 extern NTSTATUS wg_transform_read_data(void *args) DECLSPEC_HIDDEN;
 
-extern GstBufferPool *wg_pool_create(void) DECLSPEC_HIDDEN;
+extern GstBufferPool *wg_pool_create(GstCaps *caps) DECLSPEC_HIDDEN;
 extern void wg_pool_destroy(GstBufferPool *gst_pool) DECLSPEC_HIDDEN;
 
 #endif /* __WINE_WINEGSTREAMER_UNIX_PRIVATE_H */
diff --git a/dlls/winegstreamer/wg_pool.c b/dlls/winegstreamer/wg_pool.c
index 11111111111..11111111111 100644
--- a/dlls/winegstreamer/wg_pool.c
+++ b/dlls/winegstreamer/wg_pool.c
@@ -29,6 +29,7 @@
 
 #include <gst/gst.h>
 #include <gst/video/video.h>
+#include <gst/audio/audio.h>
 
 #include "unix_private.h"
 
@@ -38,6 +39,12 @@ GST_DEBUG_CATEGORY_EXTERN(wine);
 typedef struct
 {
     GstVideoBufferPool parent;
+    bool is_video;
+    union
+    {
+        GstVideoInfo video_info;
+        GstAudioInfo audio_info;
+    } u;
 } WgBufferPool;
 
 typedef struct
@@ -90,8 +97,9 @@ static void wg_pool_class_init(WgBufferPoolClass *klass)
     root_class->finalize = wg_pool_finalize;
 }
 
-GstBufferPool *wg_pool_create(void)
+GstBufferPool *wg_pool_create(GstCaps *caps)
 {
+    const gchar *media_type = gst_structure_get_name(gst_caps_get_structure(caps, 0));
     GstBufferPool *gst_pool;
     WgBufferPool *pool;
 
@@ -99,8 +107,18 @@ GstBufferPool *wg_pool_create(void)
         return NULL;
     gst_pool = GST_BUFFER_POOL(pool);
 
+    pool->is_video = g_str_has_prefix(media_type, "video/");
+    if (!(pool->is_video ? gst_video_info_from_caps(&pool->u.video_info, caps)
+            : gst_audio_info_from_caps(&pool->u.audio_info, caps)))
+        goto failed;
+
     GST_LOG("Created buffer pool %p", pool);
     return gst_pool;
+
+failed:
+    GST_ERROR("Failed to create buffer pool");
+    g_object_unref(pool);
+    return NULL;
 }
 
 void wg_pool_destroy(GstBufferPool *gst_pool)
diff --git a/dlls/winegstreamer/wg_transform.c b/dlls/winegstreamer/wg_transform.c
index 11111111111..11111111111 100644
--- a/dlls/winegstreamer/wg_transform.c
+++ b/dlls/winegstreamer/wg_transform.c
@@ -244,7 +244,7 @@ NTSTATUS wg_transform_create(void *args)
 
     if (!(transform->sink_caps = wg_format_to_caps(&output_format)))
         goto out;
-    if (!(transform->pool = wg_pool_create()))
+    if (!(transform->pool = wg_pool_create(transform->sink_caps)))
         goto out;
     if (!(sink_caps = gst_caps_copy(transform->sink_caps)))
         goto out;

